addons/addon-environment-sc-api/packages/environment-sc-workflow-steps/lib/steps/start-ec2-environment/start-ec2-environment-sc.js

```diff
@@ -0,0 +1,186 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License").
+ *  You may not use this file except in compliance with the License.
+ *  A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+const _ = require('lodash');
+const StepBase = require('@aws-ee/base-workflow-core/lib/workflow/helpers/step-base');
+
+class StartEc2EnvironmentSc extends StepBase {
+  async start() {
+    const environmentId = await this.payload.string('environmentId');
+    this.state.setKey('STATE_ENVIRONMENT_ID', environmentId);
+
+    const requestContext = await this.payload.object('requestContext');
+    this.state.setKey('STATE_REQUEST_CONTEXT', requestContext);
+
+    const instanceId = await this.payload.string('instanceIdentifier');
+
+    const ec2 = await this.getEc2Service();
+
+    const params = {
+      InstanceIds: [instanceId],
+    };
+
+    let instanceStatusInfo;
+    try {
+      const paramsForDescribe = { ...params, IncludeAllInstances: true };
+      instanceStatusInfo = await ec2.describeInstanceStatus(paramsForDescribe).promise();
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error('describe ec2 instance error: ', error);
+      throw error;
+    }
+
+    const status = _.get(instanceStatusInfo, 'InstanceStatuses[0].InstanceState.Name').toUpperCase();
+
+    if (status !== 'STOPPED') {
```

Comment:
The API is idempotent now since the environment is "locked" in STARTING status in DDB table before the workflow is invoked. Other requests coming in won't be able to trigger the workflow since they will check DDB first, see it in STARTING status, and errors out. The only possibility here is the instance was started outside of SWB and it is out of sync with DDB table. I'm not sure if the logic should try to update the DDB table status to RUNNING if that's the case? (Apologize if it does not make a whole lot of sense. We can do a quick call if needed : )

Reply:
Synced up over a call. I didn't realize we were also updating the DynamoDB status here. My main comment was around not executing `startInstance` and `checkInstanceStarted` methods since the instance was already in `start ` state. We can update DynamoDB directly.